### DISCUSSION TITLE
Debug Pode

### DISCUSSION THREAD

**Discussion by stiezinger** [Badgerati/Pode#783] [Q&A] [answered]
Hello, I am using Pode as an REST Api and it is great! Thank you for sharing this wonderfult project.
It would be perfect if I could debug some things, set breakpoints and view the values on each request.

I am using Pode v2.2.3, Visual Studio Code 1.57.1 with the PowerShell extension, ms-vscode v2021.2.2.
Hosting it with IIS. It seems I am doing something wrong because none of my breakpoints get triggered.

Here is a sample code that I use to test the breakpoint on line `$DebugVar = 1 + 1;`

```
Start-PodeServer {
    
    # IIS binding http 127.0.0.1:5050
    Add-PodeEndpoint -Address localhost -Protocol Http

    Add-PodeRoute -Method Get -Path '/' -ScriptBlock {
        $DebugVar = 1 + 1;
        Write-PodeJsonResponse -Value ($DebugVar | ConvertTo-Json)
    }
}

```
I would like to debug inside vsc with the config provided from the PowerShell extension, 
this configs I have already tried:

`{
    "name": "PowerShell Interactive Session",
    "type": "PowerShell",
    "request": "launch",
    "cwd": ""
}`

`{
    "name": "PowerShell: Attach to PowerShell Host Process",
    "type": "PowerShell",
    "request": "attach",
    "cwd": ""
}`

`{
    "name": "PowerShell Attach Interactive Session Runspace",
    "type": "PowerShell",
    "request": "attach",
    "cwd": ""
}`

how do you usually debug?

**Comment by Badgerati** [comment, marked as answer]
Hey @stiezinger,

Usually when I'm debugging, I just pipe stuff into `Out-Default` or use `New-PodeLoggingMethod -Terminal | Enable-PodeErrorLogging` to get values onto the terminal 😂 

Pode uses a lot of runspaces behind the scenes, which is why the breakpoints won't get triggered. It's something I've been meaning to look into properly for quite sometime, being able to use `Debug-Runspace` and document how to do it. It might be possible to somehow make VSC use it's breakpoints, but I'll need to properly investigate how to (and if anyone here has managed it, feel free to say how!).

The last time I tried looking into this, it went something like this:
* Run your Pode server (make sure to use `-Threads 1` on `Start-PodeServer`
* Open a PowerShell terminal
* Run `Get-Process powershell` and get the process ID of the process running Pode
* Run `Enter-PSHostProcess -Id <processId>`
* Run `Get-Runspace` to show all the runspaces being used by Pode, and identify the runspace you need
* Run `Debug-Runspace -Id <runspaceId>`

^ and that will let you step through the runspace when you hit it with a request. I haven't tested it properly, and I do want to look into how to make it cleaner and easier to use.

**Reply by stiezinger** [reply]
Thank you @Badgerati for the detailed answer.

**Reply by Viajaz** [reply]
PowerShell Universal seems to have an implementation but even with https://github.com/ironmansoftware/issues/issues/1284 I think the Debugging feature is an [enterprise/paid feature](https://docs.powershelluniversal.com/licensing#free-use-restrictions)

**Reply by Badgerati** [reply]
I'm not too sure about PSU, but I believe UD relied on similar `Debug-Runspace` calls while debugging. This is still on my list to try and make nicer, and document it; possibly even figure out if we can attach VSC to the powershell process and step through it 🤔

**Comment by nightroman** [comment]
Here is some peculiar but surprisingly robust bare runspace debugger:
https://www.powershellgallery.com/packages/Add-Debugger

In your Pode route handler or similar code call `Add-Debugger` and then use `Wait-Debugger` as hardcoded breakpoints or set other breakpoints by `Set-PSBreakpoint`.

When breakpoints are hit, enter debugger commands as prompted (either in console or GUI depending on the host and `Add-Debugger` parameters).

Trivial example: add the debugger and break immediately after:

```powershell
Add-Debugger
Wait-Debugger
```

**Tip**: Use `d` (Detach) in order to stop debugging instead of `q` (Quit) to avoid stopping Pode.
